lib/opte/src/engine/packet.rs

```diff
@@ -740,12 +740,22 @@ impl Packet<Initialized> {
             pkt_offset,
         );
 
-        let mut body = BodyInfo {
-            pkt_offset,
-            seg_index,
-            seg_offset,
-            len: self.state.len - pkt_offset,
+        let ulp_hdr_len = info.meta.inner.ulp.map(|u| u.hdr_len()).unwrap_or(0);
+        let body_len = match info.meta.inner.ip {
+            // If we have IP and ULP metadata, we can use those to compute
+            // the payload length.
+            // If there's no ULP, just return the L3 payload length.
+            Some(IpMeta::Ip4(ip4)) => {
+                usize::from(ip4.total_len) - ip4.hdr_len() - ulp_hdr_len
+            }
+            Some(IpMeta::Ip6(ip6)) => usize::from(ip6.pay_len) - ulp_hdr_len,
+
+            // If there's no IP metadata, we fallback to considering any
+            // remaining bytes in the packet buffer to be the body.
+            None => self.state.len - pkt_offset,
```

Comment:
perhaps this is worth a log entry?

Reply:
Of course, if we expect this will be a normal condition this will be a really hot path and logging would be bad here.

Reply:
I didn't add one here specifically because the `NetworkParser` impls will fail in the `parse_{inbound,outbound}` calls above if there's no inner IP. The only exception being for ARP packets.